### PR TITLE
DAOS-5781 test: Change the output handling in launch.py

### PR DESCRIPTION
**Pull request by ashleypittman**
Print each line of remote output individually, rather than making
into one long string.  Add a delay and retry in case of IO errors
which we are seeing occasionally.

Change the free functions used in some test code and add some
more whitelists for memory allocation facilities.

Have cart_logtest catch errors and only report to stdout rather than
propagating, and extend this to when it's called with a directory as well.

### CODE REVIEW

File: src/tests/ftest/launch.py [hunk 1]
```diff
@@ -338,10 +338,13 @@ def check_remote_output(task, command):
...
             for output, o_hosts in output_data:
                 n_set = NodeSet.fromlist(o_hosts)
-                lines = str(output).splitlines()
+                lines = output.decode('utf-8').splitlines()
                 if len(lines) > 1:
-                    output = "\n      {}".format("\n      ".join(lines))
-                print("    {}: rc={}, output: {}".format(n_set, code, output))
+                    print("    {}: rc={}, output:".format(n_set, code))
+                    for line in lines:
+                        print("      {}".format(lines)
```
Comment (ashleypittman) [inline src/tests/ftest/launch.py:345]:
```suggestion
                        print("      {}".format(lines))
```

File: src/tests/ftest/launch.py [hunk 2]
```diff
@@ -338,10 +338,13 @@ def check_remote_output(task, command):
...
             for output, o_hosts in output_data:
                 n_set = NodeSet.fromlist(o_hosts)
-                lines = str(output).splitlines()
+                lines = output.decode('utf-8').splitlines()
                 if len(lines) > 1:
-                    output = "\n      {}".format("\n      ".join(lines))
-                print("    {}: rc={}, output: {}".format(n_set, code, output))
+                    print("    {}: rc={}, output:".format(n_set, code))
+                    for line in lines:
+                        print("      {}".format(lines))
```
Comment (ashleypittman) [inline src/tests/ftest/launch.py:345]:
```suggestion
                        print("      {}".format(line))
```

File: src/tests/ftest/launch.py [hunk 3]
```diff
@@ -339,9 +339,20 @@ def check_remote_output(task, command):
...
-                    output = "\n      {}".format("\n      ".join(lines))
-                print("    {}: rc={}, output: {}".format(n_set, code, output))
+                    print("    {}: rc={}, output:".format(n_set, code))
+                    for line in lines:
+                        try:
+                            print("      {}".format(line))
+                        except IOError:
+                            time.sleep(5)
+                            print('XXX: IO Error')
+                            print("      {}".format(line))
```
Comment (mjmac) [inline src/tests/ftest/launch.py:351]: What problem is this intended to work around? Some kind of buffering issue on stdout?

Comment (ashleypittman) [reply src/tests/ftest/launch.py]: Yes, that's my theory.

It matches as well, this PR passed and we did see the debug line.  The test runs a bunch of stuff by ssh then batches all it's output so it must just be filling a buffer in Jenkins itself somewhere.

The leap build passed on this commit but I've just pushed another one now which removes the debug print and whitelists some more memory locations, hopefully it'll have passed by tomorrow.

**Review by sylviachanoiyee — APPROVED** [review]

**Review by mjmac — APPROVED** [review]
LGTM

**Pull request merged**